src/MessagePassing/clustercorrections.jl

```diff
@@ -0,0 +1,216 @@
+                current_weight,
+                length(current_cluster)
+            )
+            
+            # Avoid duplicates using canonical signature
+            signature = canonical_cluster_signature(cluster)
+            if !(signature in seen_clusters)
+                push!(seen_clusters, signature)
+                push!(clusters, cluster)
+                cluster_count += 1                
+            end
+        end
+        
+        # Stop if we've reached max weight
+        if current_weight >= max_weight
+            return
+        end
+        
+        # Find candidate loops to add (adjacent loops or multiplicities)
+        candidate_loops = Set{Int}()
+        
+        if isempty(current_cluster)
+            # Start with supported loops only
+            for loop_id in supported_loop_ids
+                if all_loops[loop_id].weight <= max_weight - current_weight
+                    push!(candidate_loops, loop_id)
+                end
+            end
+        else
+            # Add loops connected to current cluster via interaction graph
+            for loop_id in current_cluster
+                # Add connected loops (touching loops)
+                for neighbor_id in get(interaction_graph, loop_id, Int[])
+                    if all_loops[neighbor_id].weight <= max_weight - current_weight
+                        push!(candidate_loops, neighbor_id)
+                    end
+                end
+                # Allow multiplicity increases (same loop added again)
+                if all_loops[loop_id].weight <= max_weight - current_weight
+                    push!(candidate_loops, loop_id)
+                end
+            end
+        end
+        
+        # Try each candidate loop
+        for loop_id in candidate_loops
+            loop_weight = all_loops[loop_id].weight
+            new_weight = current_weight + loop_weight
+            
+            if new_weight <= max_weight
+                new_cluster = copy(current_cluster)
+                push!(new_cluster, loop_id)
+                new_has_supported = has_supported || (loop_id in supported_loop_ids)
+                
+                # Continue DFS (connectivity guaranteed by interaction graph)
+                dfs_grow_cluster(new_cluster, new_weight, new_has_supported)
+            end
+        end
+    end
+    
+    # Start DFS with empty cluster
+    dfs_grow_cluster(Int[], 0, false)
+    
+    verbose && println("  DFS enumeration completed: $cluster_count total clusters found")
+    return clusters
+end
+
+"""
+Build all clusters on named graph ng, up to a given weight. Optionally, must be supported on the vertices must_contain, in which case those vertices can be leaves
+
+This is overkill as it finds ALL subgraphs first, but my other implementation had bugs
+"""
+function enumerate_clusters(ng::NamedGraph, max_weight::Int; min_v::Int = 4, triangle_free::Bool = true, must_contain = [], min_deg::Int = 2, verbose::Bool = false)
```

Comment:
So this should definitely return the clusters with vertex labelling that `ng` has.

This is how I do it for `edge_color` (basically define the function on a `SimpleGraph` and then overload it on the `NamedGraph` by unwrapping the graph and re-wrapping the result:

```
function SimpleGraphAlgorithms.edge_color(g::AbstractNamedGraph, k::Int64)
  pg, vs = position_graph(g), collect(vertices(g))
  ec_dict = SimpleGraphAlgorithms.edge_color(UndirectedGraph(pg), k)
  # returns k vectors of edges which each contain the colored/commuting edges
  return [
    [edgetype(g)(vs[first(first(e))], vs[last(first(e))]) for e in ec_dict if last(e) == i]
    for i in 1:k
  ]
end
end
```